umbra-js/src/utils/utils.ts

```diff
@@ -15,40 +24,42 @@ const constants = require('../constants.json');
  *      strip leading zeroes.
  *   3. When generating random numbers and returning them as hex strings, the leading
  *      zero bytes get stripped
- * @param {String} hex String to pad, without leading 0x
- * @param {String} bytes Number of bytes string should have
+ * @param hex String to pad, without leading 0x
+ * @param bytes Number of bytes string should have
  */
-module.exports.padHex = (hex, bytes = 32) => {
-  if (!utils.isHexString) throw new Error('Input is not a valid hex string');
+export function padHex(hex: string, bytes = 32) {
+  if (!isHexString) throw new Error('Input is not a valid hex string');
   if (hex.slice(0, 2) === '0x') throw new Error('Input must not contain 0x prefix');
-  return hex.padStart(bytes * 2, 0);
-};
+  return hex.padStart(bytes * 2, '0');
+}
 
 /**
  * @notice Convert hex string with 0x prefix into Buffer
- * @param {String} data Hex string to convert
+ * @param value Hex string to convert
  */
-module.exports.hexStringToBuffer = (data) => Buffer.from(utils.arrayify(data));
+export function hexStringToBuffer(value: string | number | Bytes | Hexable) {
+  return Buffer.from(arrayify(value));
+}
 
 /**
  * @notice Given a transaction hash, return the public key of the transaction's sender
  * @dev See https://github.com/ethers-io/ethers.js/issues/700 for an example of
  * recovering public key from a transaction with ethers
  * @param {String} txHash Transaction hash to recover public key from
- * @param {*} provider raw web3 provider to use (not an ethers instance)
+ * @param {*} provider web3 provider to use (not an ethers provider)
  */
-module.exports.recoverPublicKeyFromTransaction = async (txHash, provider) => {
+export async function recoverPublicKeyFromTransaction(txHash: string, provider: ExternalProvider) {
   // Get transaction data
-  const ethersProvider = new ethers.providers.Web3Provider(provider);
+  const ethersProvider = new Web3Provider(provider);
   const tx = await ethersProvider.getTransaction(txHash);
 
   // Get original signature
-  const splitSignature = {
-    r: tx.r,
+  const splitSignature: SignatureLike = {
+    r: tx.r as string,
```

Comment:
Why does this have to be cast as a string here? Doesn't Ethers have full type definitions? I tried looking up the `getTransaction` method in the Ethers doc, but couldn't find it. Weird.

Reply:
`tx` is of type [`TransactionResponse`](https://github.com/ethers-io/ethers.js/blob/efeeda5e5c3b22f0c211783a635941dc8035ae25/packages/abstract-provider/src.ts/index.ts#L31-L49), which extends [`Transaction`](https://github.com/ethers-io/ethers.js/blob/efeeda5e5c3b22f0c211783a635941dc8035ae25/packages/transactions/src.ts/index.ts#L31-L48).  The `getTransaction` method is [here](https://github.com/ethers-io/ethers.js/blob/8933467c01b64ead547d7c136f22f3c05c85ca1f/packages/providers/src.ts/base-provider.ts#L1183-L1215) (but like you said, for some reason I can't find it in the docs!)

So as you can see the `r,s,v` components of `Transaction` don't necessarily exist (since that is used to send transactions also), and since `TransactionResponse` extends that, TS thinks the signature components may be undefined we cast to a string. The other two don't need to be casted because they're optional in `SignatureLike`

However, this is a good point. I think the `r,s,v` components should always exist in `TransactionResponse`, so I'll create an issue in ethers to ask about this

Reply:
we could do a check above: `if(!tx.r) throw new Error("Invalid signature data");` then we can avoid casting to string here. up to you guys.